ludwig/hyperopt/utils.py

```diff
@@ -141,3 +173,105 @@ def substitute_parameters(
     config = feature_dict_to_list(config)
 
     return config
+
+
+def log_warning_if_all_grid_type_parameters(hyperopt_parameter_config: Dict[str, Any], num_samples: int = 1) -> None:
+    """Logs warning if all parameters have a grid type search space and num_samples > 1 since this will result in
+    duplicate trials being created."""
+    if num_samples == 1:
+        return
+
+    total_grid_search_trials = 1
+
+    for _, param_info in hyperopt_parameter_config.items():
+        if param_info.get(SPACE, None) != GRID_SEARCH:
+            return
+        total_grid_search_trials *= len(param_info.get("values", []))
+
+    num_duplicate_trials = (total_grid_search_trials * num_samples) - total_grid_search_trials
+    warnings.warn(
+        "All hyperopt parameters in Ludwig config are using grid_search space, but number of samples "
+        f"({num_samples}) is greater than 1. This will result in {num_duplicate_trials} duplicate trials being "
+        "created. Consider setting `num_samples` to 1 in the hyperopt executor to prevent trial duplication.",
+        RuntimeWarning,
+    )
+
+
+def get_total_trial_count(hyperopt_parameter_config: Dict[str, Any], num_samples: int = 1) -> int:
+    """Returns the total number of hyperopt trials that will run based on the hyperopt config."""
+    total_trial_count = num_samples
+    for _, param_info in hyperopt_parameter_config.items():
+        if param_info.get(SPACE) == GRID_SEARCH:
+            total_trial_count *= len(param_info.get("values"))
+    return total_trial_count
+
+
+def update_hyperopt_params_with_defaults(hyperopt_params: Dict[str, Any]) -> None:
+    """Updates user's Ludwig config with default hyperopt parameters."""
+    from ludwig.hyperopt.execution import executor_registry
+
+    set_default_value(hyperopt_params, EXECUTOR, {})
+    set_default_value(hyperopt_params, SPLIT, VALIDATION)
+    set_default_value(hyperopt_params, "output_feature", COMBINED)
+    set_default_value(hyperopt_params, METRIC, LOSS)
+    set_default_value(hyperopt_params, GOAL, MINIMIZE)
+
+    set_default_values(hyperopt_params[EXECUTOR], {TYPE: RAY, NUM_SAMPLES: 1, MAX_CONCURRENT_TRIALS: AUTO})
+
+    executor = get_from_registry(hyperopt_params[EXECUTOR][TYPE], executor_registry)
+    executor_defaults = {k: v for k, v in executor.__dict__.items() if k in get_class_attributes(executor)}
+    set_default_values(
+        hyperopt_params[EXECUTOR],
+        executor_defaults,
+    )
+
+
+def update_or_set_max_concurrent_trials(hyperopt_config: dict, backend: Backend) -> None:
```

Comment:
Let's add unit (not integration) tests for this function. Can put them in `tests/ludwig/backend.ray.py` if we make the change I suggested to move this into the Backend.

Reply:
Looks like we can just take the unit tests from `test_hyperopt.py` and adapt them to the new API.